src/librbd/Utils.h

```diff
@@ -46,6 +46,25 @@ void rados_state_callback(rados_completion_t c, void *arg) {
   }
 }
 
+template <typename T>
+class C_ContextCallbackAdapter : public Context {
+  T *obj;
```

Comment:
Comment was lost in the previous review:

I don't see why this should be a template. Why not just use `RefCountedObjectPtr obj` instead? Then you don't need to do the manual ref count changes using `->get()`/`->put()`.

Reply:
> I don't see why this should be a template. Why not just use `RefCountedObjectPtr obj` instead? Then you don't need to do the manual ref count changes using `->get()`/`->put()`.

Do you mean `RefCountedObject *obj` ? This shouldn't have to be a template and could instead use `RefCountedObject *obj`. But I'd still have to call `get()` and `put()` since the objects are not created using ceph::ref_t (boost::intrusive_ptr).

Reply:
No. It should be as simple as:

````
class Foo ... {
  RefCountedObjPtr obj;
  public:  Foo(RefCountedObjectPtr o, ...) : obj(std::move(o)) ... {}
};
````

>  But I'd still have to call get() and put() since the objects are not created using ceph::ref_t (boost::intrusive_ptr).

That doesn't matter. They are still reference counted. The coercion of a `RefCountedObject*` to `ceph::ref_t` will automatically call `->get()` for you. And of course, when the `ceph::ref_t` is destructed, `->put()` will also be called.

Reply:
Ah okay. I couldn't find `RefCountedObjPtr` in the code. But I now understand, you mean this `RefCountedPtr`. Thanks!